src/game/client/gui/guicallbacks/controlbarcallback.cpp

```diff
@@ -12,18 +12,360 @@
+{
+    return MSG_IGNORED;
+}
+
+WindowMsgHandledType Control_Bar_System(GameWindow *window, unsigned int message, unsigned int data_1, unsigned int data_2)
+{
+    GameWindow *window2 = reinterpret_cast<GameWindow *>(data_1);
+    static NameKeyType buttonCommunicator;
+
+    if (g_theScriptEngine != nullptr && g_theScriptEngine->Is_End_Game_Timer_Running()) {
+        return MSG_IGNORED;
+    }
+
+    if (message > GBM_SELECTED_RIGHT) {
+        if (message == GEM_EDIT_DONE) {
+            int id = window2->Win_Get_Window_Id();
+            static const NameKeyType textID = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:EditBeaconText");
+
+            if (id == textID && g_theInGameUI->Get_Select_Count() == 1) {
+                GameMessage *msg = g_theMessageStream->Append_Message(GameMessage::MSG_SET_BEACON_TEXT);
+                Utf16String text = Gadget_Text_Entry_Get_Text(window2);
+                g_theLanguageFilter->Filter_Line(text);
+                const unichar_t *str = text.Str();
+
+                while (str != nullptr && *str != '\0') {
+                    msg->Append_Wide_Char_Arg(*str);
+                    str++;
+                }
+
+                msg->Append_Wide_Char_Arg('\0');
+            }
+
+            return MSG_HANDLED;
+        }
+
+        return MSG_IGNORED;
+    }
+
+    if (message >= GBM_SELECTED) {
+        static const NameKeyType beaconPlacementButtonID =
+            g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonPlaceBeacon");
+        static const NameKeyType beaconDeleteButtonID =
+            g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonDeleteBeacon");
+        static const NameKeyType beaconClearTextButtonID =
+            g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonClearBeaconText");
+        static const NameKeyType beaconGeneralButtonID = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonGeneral");
+        static const NameKeyType buttonLargeID = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonLarge");
+        static const NameKeyType buttonOptions = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonOptions");
+        static const NameKeyType buttonIdleWorker = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonIdleWorker");
+        int id = window2->Win_Get_Window_Id();
+
+        if (id == buttonCommunicator) {
+            Toggle_Diplomacy(false);
+        } else if (id == beaconPlacementButtonID && g_theGameLogic->Is_In_Multiplayer_Game()
+            && g_thePlayerList->Get_Local_Player()->Is_Player_Active()) {
+            g_theInGameUI->Set_GUI_Command(g_theControlBar->Find_Command_Button("Command_PlaceBeacon"));
+        } else if (id == beaconDeleteButtonID && g_theGameLogic->Is_In_Multiplayer_Game()) {
+            g_theMessageStream->Append_Message(GameMessage::MSG_REMOVE_BEACON);
+        } else if (id == beaconClearTextButtonID && g_theGameLogic->Is_In_Multiplayer_Game()) {
```

Comment:
`g_theGameLogic->Is_In_Multiplayer_Game()` is not a sub condition? Seems odd that it tests on if not in multiplayer but the button id was matched.

Several times.

Reply:
This is correct.